Commands for CMake

iguessthislldo opened the pull request:
Added `cmake`, `print_cmake_version`, and `cmake_cmd` commands, new `<arg>` syntax, and tests for the new changes. Simplest use of `cmake` command is:

```xml
<command name="cmake" dir="the_project"/>
```

Where `the_project` is the directory with the `CMakeLists.txt` file to build. Default parameters can be changed using new `<arg>` tags. This is equivalent to the command above:

```xml
<command name="cmake" dir="the_project">
  <arg name="build_dir">build</arg>
  <arg name="config_args">..</arg>
  <arg name="build_args">--build .</arg>
</command>
```

Also added a separate command to invoke CMake directly:

```xml
<command name="cmake_cmd" options="-E capabilities"/>
```

Also changed the `file_manipulation` command to support the new `<arg>` tags while keeping it compatible with the previous behavior. These two commands are equivalent:

```xml
<command name="file_manipulation" options="type=create file=file.txt">
 <arg name="output">
if (x &gt; 5)
 print("Greater than 5");
 </arg>
</command>
```

```xml
<command name="file_manipulation"  options="type=create file=file.txt">
if (x &gt; 5)
  print("Greater than 5");
</command>
```

iguessthislldo commented:
I've reopened this because I would like to be able to use these commands now. I don't completely remember why I closed it. I think it was that I didn't like you manually to change directories to and back as show in the example above, where as with anonymous shell you can cd in each command and it will automatically come back. I don't think I can simplify any more though, so I'm fine with how it is now.

iguessthislldo commented:
Alternatively I could merge all these actictions:

```xml
  <command name="file_manipulation" options="type=rmmkcd file=build"/>
  <command name="cmake" options=".."/>
  <command name="cmake" options="--build ."/>
  <command name="file_manipulation" options="type=chdir"/>
```
into the cmake command, so it would look something like:

```xml
  <command name="cmake" options="replace=build .."/>
```

mitza-oci commented:
> 
> 
> Alternatively I could merge all these actictions:
> 
> ```
>   <command name="file_manipulation" options="type=rmmkcd file=build"/>
>   <command name="cmake" options=".."/>
>   <command name="cmake" options="--build ."/>
>   <command name="file_manipulation" options="type=chdir"/>
> ```
> 
> into the cmake command, so it would look something like:
> 
> ```
>   <command name="cmake" options="replace=build .."/>
> ```

I like the alternative you've shown here.  That seems to fit with the way most of the autobuild commands work.

iguessthislldo commented:
The only issue I have with that approach is that I would like to address three different kinds of options: the output directory name/location, the arguments to `cmake` and the arguments to `cmake --build` and I would like to be able to define them all, but I'm not sure how to that.

mitza-oci commented:
> The only issue I have with that approach is that I would like to address three different kinds of options: the output directory name/location, the arguments to `cmake` and the arguments to `cmake --build` and I would like to be able to define them all, but I'm not sure how to that.

Many autobuild commands use `<variable>` XML elements to set additional parameters.

Or how about something like this, a hybrid of the options you discussed before:
```
  <command name="file_manipulation" options="type=rmmk file=build"/>
  <command name="cmake" options="dir=build .."/>
  <command name="cmake" options="dir=build --build ."/>
```

iguessthislldo commented:
Another idea is to extend the parser to accept tags inside the command tag like so:

```xml
<command name="cmake"/>
  <opt name="dir">build_dir_name</option>
  <opt name="config_args">--an-arg-for-config ..</opt>
  <opt name="build_args">--another-arg-for-build</opt>
</command>
```

This would have to resolved with #19, so I'm thinking it could be generalized that values inside the `options` attribute of `command` could be also be passed as separate `opt` tags:

```xml
<command name="file_manipulation"  options="type=create file=setenv.sh">
  <opt name="output">
export ACE_ROOT="${ACE_WORKSPACE}/ATCD/ACE"
export CIAO_ROOT="${ACE_WORKSPACE}/ATCD/CIAO"
export DANCE_ROOT="${ACE_WORKSPACE}/ATCD/DAnCE"
export DDS_ROOT="${WORKSPACE}/OpenDDS"
export LD_LIBRARY_PATH="/usr/sfw/lib:/usr/local/lib:${WORKSPACE}/OpenDDS/lib:${ACE_WORKSPACE}/ATCD/ACE/lib:"
export MPC_ROOT="/home/taoadmin/MPC"
export OPENDDS_RTPS_DEFAULT_D0="146"
export PATH="${WORKSPACE}/OpenDDS/bin:${ACE_WORKSPACE}/ATCD/ACE/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/ccs/bin:"
export TAO_ROOT="${ACE_WORKSPACE}/ATCD/TAO"
  </opt>
</command>

mitza-oci commented:
That seems like more than you really need.  Can it parse XML Attributes?
```
<command name="cmake" dir="build_dir_name" config_args="args1" build_args="args2"/>
```

mitza-oci commented on command/file_manipulation.pm:
```diff
@@ -853,7 +853,7 @@ sub Run ($)
         return 0;
     }
-    if (!defined ($filename) || ($filename eq "")) {
+    if ($type ne 'chdir' && (!defined ($filename) || ($filename eq ""))) {
```
Maybe cleaner just to have the 'chdir' handling first?

mitza-oci commented on command/file_manipulation.pm:
```diff
@@ -1116,6 +1116,41 @@ sub Run ($)
+        if (defined $filename) {
+            if (-d $filename ) {
+                $current_dir = $filename;
+            } else {
```
```suggestion
            }
            else {
```

mitza-oci commented on command/file_manipulation.pm:
```diff
@@ -1116,6 +1116,41 @@ sub Run ($)
+            $current_dir = $filename;
+        } else {
+            print STDERR __FILE__, ":\n",
+                "  Could not chdir into $filename, it is not a directory!\n";
```
Should this be an error about `mkdir`?
Is there a missing `chdir` in the sequence?

iguessthislldo replied:
There is a preexisting `chdir` at the end of the function that this command and the `chdir` command are using.

I could also remove this command, since I'm not planning on using it anymore.

iguessthislldo commented:
> That seems like more than you really need. Can it parse XML Attributes?

Yes, but they'd be common to all `command` tags, so I don't think that's the right way to extend this. There is already a [`dir` (with other names) attribute](https://github.com/DOCGroup/autobuild/blob/540a5519ee6e016a321f19d9c2af68a3ace0a604/common/betterparser.pm#L1257). I'd have to see what it does now, but it makes sense to reuse that if possible. For the cmake specific stuff, I checked the `<opt>` tag code I have locally that I haven't push is almost done, just needs to be cleaned up. I really think it's the way to go instead of continuing to shove things inside the attributes.

iguessthislldo commented:
So I realized `chdir` isn't necessary in other commands other than `cmake` because the `dir` command attribute seems to change the directory in a lot of commands. `chdir` command also doesn't work with these other commands how it is now because the commands are calling `chdir($root)` at the start to support the `dir` attribute. I thought it was working because it was working with `cmake`, but that's because I don't have that `chdir` call.

So I'm going to remove `chdir` in addition to `rmmkcd` and have `cmake` support `dir`.

iguessthislldo commented:
Okay so I took the existing work I had locally and cleaned it up. I also renamed `opt` tags to `arg` to attempt to avoid confusion with the `options` attribute, though I'm not sure how much it helps. Using the `dir` attribute, the most complicated cmake command looks like this:

```xml
<command name="cmake" dir="the_project">
  <arg name="build_dir">build</arg>
  <arg name="config_args">..</arg>
  <arg name="build_args">--build .</arg>
</command>
```

But if using all the default args like above, it can just be this:

```xml
<command name="cmake" dir="the_project"/>
```

Also added a separate command to invoke CMake directly:

```xml
<command name="cmake_cmd" options="-E capabilities"/>
```

Unlike what I previously stated over a year ago, I was able to keep `file_manipulation` backwards compatible and make it able to accept the new `<arg>` tags, so the following two are equivalent:

```xml
<command name="file_manipulation" options="type=create file=file.txt">
 <arg name="output">
if (x &gt; 5)
 print("Greater than 5");
 </arg>
</command>
```

```xml
<command name="file_manipulation"  options="type=create file=file.txt">
if (x &gt; 5)
  print("Greater than 5");
</command>
```

This allows new args to be added to ` file_manipulation` in the future.

Finally I added tests and a workflow to run them. They are fairly basic, but at least it's something. I was thinking we could also add simple scoreboard tests since we've been having trouble with that lately.

iguessthislldo commented on command/cmake.pm:
```diff
@@ -0,0 +1,150 @@
+        return 0;
+    }
+
+    {
```
Hmmm, I want to say it's for `ChangeDir` because it changes back when the object is done, but it returns after this, so no it's not needed.
